module1/homework/grayscaleImages/compression.cpp

```diff
@@ -0,0 +1,67 @@
+#include "compression.hpp"
+
+#include <iostream>
+#include <string>
+
+std::vector<std::pair<uint8_t, uint8_t>> compressGrayscale(
+    std::array<std::array<uint8_t, height>, width>& uncompressed) {
+    std::vector<std::pair<uint8_t, uint8_t>> compressed;
+    compressed.reserve(height * width);
+
+    for (const auto& row : uncompressed) {
+        auto rowIterator = row.cbegin();
+        uint8_t value = *rowIterator;
+        uint8_t currentValueCounter = 1;
+        rowIterator++;
+
+        while (rowIterator != row.cend()) {
+            if (*rowIterator == value) {
+                currentValueCounter++;
+            } else {
+                compressed.push_back(std::make_pair(value, currentValueCounter));
+                value = *rowIterator;
+                currentValueCounter = 1;
+            }
+            rowIterator++;
+        }
+        // and for last value in a row:
+        compressed.push_back(std::make_pair(value, currentValueCounter));
+    }
+
+    compressed.shrink_to_fit();
+    return compressed;
+}
+
+std::array<std::array<uint8_t, height>, width> decompressGrayscale(
+    std::vector<std::pair<uint8_t, uint8_t>>& compressed) {
+    std::array<std::array<uint8_t, height>, width> uncompressed;
+    auto pixelGroup = compressed.begin();
+    uint8_t howManyPixels = pixelGroup->second;
+
+    for (auto& row : uncompressed) {
+        for (auto& pixel : row) {
+            if (howManyPixels == 0) {
+                pixelGroup++;
+                howManyPixels = pixelGroup->second;
+            }
+            pixel = pixelGroup->first;
+            howManyPixels--;
```

Comment:
Decrement value `howManyPixel` is misleading. Try to use normal loop.

Reply:
You could say that, but because he uses temporary variable to store number of pixels and iterates over rows and columns he protected himself from invalid output that most naive implementatiions just naively accept causing UB because of the overflow.
You can simplify it to normal loops - but you can't do it without using temporary howManyPixels as you suggest above - unless modifying pixelGroup->second is accepted.

Reply:
In newest version I've renamed this variable, so the code looks less misleading.